### PR TITLE
Implement `compute_speed` and `compute_path_length`

### LINKED ISSUE #147: Consider adding a distance and speed property to accessor
From the kinematics tutorial PR review:

We could add a "distance" property in our accessor to access directly the distance travelled, e.g:
```python
distance_travelled = ds.move.distance
```
Similarly with speed.

### PR DESCRIPTION
**Pull request by niksirbi**
## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**What does this PR do?**

It introduces two new functions in the kinematics module:
- `compute_speed`: quite straightforward to implement, it just returns the norm of velocity, using existing `movement` functions
- `compute_path_length`: in theory also straightforward, as it should just be the sum of the norms of displacement vectors, but I spent a lot of time debating how to handle missing values. See extensive [discussion on zulip](https://neuroinformatics.zulipchat.com/#narrow/stream/406001-Movement/topic/Path.20length.20with.20missing.20values).

In the end I opted for providing two alternative nan policies, and emitting a warning if too many values are missing. Expand the dropdown below for more details.

<details>
<summary> compute_path_length() function signatures and docstring</summary>

```python
def compute_path_length(
    data: xr.DataArray,
    start: float | None = None,
    stop: float | None = None,
    nan_policy: Literal["drop", "scale"] = "drop",
    nan_warn_threshold: float = 0.2,
) -> xr.DataArray:
    """Compute the length of a path travelled between two time points.

    The path length is defined as the sum of the norms (magnitudes) of the
    displacement vectors between two time points ``start`` and ``stop``,
    which should be provided in the time units of the data array.
    If not specified, the minimum and maximum time coordinates of the data
    array are used as start and stop times, respectively.

    Parameters
    ----------
    data : xarray.DataArray
        The input data containing position information in Cartesian
        coordinates, with ``time`` and ``space`` among the dimensions.
    start : float, optional
        The time to consider as the path's starting point. If None (default),
        the minimum time coordinate in the data is used.
    stop : float, optional
        The time to consider as the path's end point. If None (default),
        the maximum time coordinate in the data is used.
    nan_policy : Literal["drop", "scale"], optional
        Policy to handle NaN (missing) values. Can be one of the ``"drop"``
        or ``"scale"``. Defaults to ``"drop"``. See Notes for more details
        on the two policies.
    nan_warn_threshold : float, optional
        If more than this proportion of values are missing in any point track,
        a warning will be emitted. Defaults to 0.2 (20%).

    Returns
    -------
    xarray.DataArray
        An xarray DataArray containing the computed path length.
        Will have the same dimensions as the input data, except for ``time``
        and ``space`` which will be removed.

    Notes
    -----
    Choosing ``nan_policy="drop"`` will drop NaN values from each point track
    before computing path length. This equates to assuming that a track
    follows a straight line between two valid points flanking a missing
    segment. Missing segments at the beginning or end of the specified
    time range are not counted. This approach tends to underestimate
    the path length, and the error increases with the number of missing
    values.

    Choosing ``nan_policy="scale"`` will adjust the path length based on the
    the proportion of valid segments per point track. For example, if only
    80% of segments are present, the path length will be computed based on
    these and the result will be divided by 0.8. This approach assumes
    that motion dynamics are similar across present and missing time
    segments, which may not be the case.

    """
```

</details>

I'm in two minds as to whether we need the `start` and `stop` arguments, as the user could easily select a time range using `ds.position.sel(time=slice(0, 10))`, before passing the data array to `compute_path_length()`. I've chosen to include them for now, but I'm open to counter-arguments, CC @b-peri.

## References

Closes #147 

## How has this PR been tested?

For speed, I used the existing kinematics tests (added "speed" as a parameter) with a bit of modification needed.

Testing for the path length was harder because:
- this variable lacks both time and space dimensions, so modifying existing kinematics tests to account for it would b complicated
- With the two aforementioned nan_policies, there are several tricky edge cases, and I tried to cover them all. The uniform linear motion fixture has been instrumental in helping me implement , debug and test these nan policies.
- I also had to test that the nan warning is emitted in the right scenarios and contains specific messages.

As a result I ended up adding 3 new unit tests to account for all of the above, but suggestion to streamline the tests are welcome.

## Is this a breaking change?

No.

## Does this PR require an update to the documentation?

API docs should be automatically updated. We might consider mentioning the new functions in some of the examples in the future, but not necessary right now.

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality
- [x] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

**Review by lochhh — CHANGES_REQUESTED** [review]
Thanks @niksirbi ! 
To summarise:
- a few suggestions for wordings, which you're free to take or leave as usual
- start, stop checks: I'm wondering if we should explicitly check these for slicing. While the checks are much more restrictive than xarray, do we need to do so? xarray, for instance, automatically constraints the slicing to the min, max, regardless of whether the provided start, stop values are within or out of range. if we drop our own checks, the relevant tests can then be much simpler
- simplify `_compute_path_length_drop_nan`: hacking with `ffill` and let xarray's `groupby` take care of stacking - needs further testing? 
- simplify tests (reduce parametrisation, test only `id_1` values in nan test)
- resolve merge conflicts with main branch

File: movement/analysis/kinematics.py [hunk 1]
```diff
@@ -343,3 +444,180 @@ def _validate_type_data_array(data: xr.DataArray) -> None:
...
+    ----------
+    data : xarray.DataArray
+        The input data array.
+    nan_warn_threshold : float
+        The threshold for the proportion of NaN values. Must be a number
+        between 0 and 1.
+
+    """
+    nan_warn_threshold = float(nan_warn_threshold)
+    if nan_warn_threshold < 0 or nan_warn_threshold > 1:
```
[the remark this answers is not in the document]
Comment (niksirbi) [reply movement/analysis/kinematics.py:526]: neat

File: tests/test_unit/test_kinematics.py [hunk 2]
```diff
@@ -186,6 +204,223 @@ def test_approximate_derivative_with_invalid_order(order):
...
+            position, start=start, stop=stop, nan_policy=nan_policy
+        )
+
+        # Expected number of segments (displacements) in selected time range
+        num_segments = 9  # full time range: 10 frames - 1
+        if start is not None:
+            num_segments -= np.ceil(start)
+        if stop is not None:
+            num_segments -= 9 - np.floor(stop)
+        print("num_segments", num_segments)
```
[the remark this answers is not in the document]
Comment (niksirbi) [reply tests/test_unit/test_kinematics.py:292]: Good catch, I'd forgotten to delete that.

File: movement/analysis/kinematics.py [hunk 3]
```diff
@@ -343,3 +444,180 @@ def _validate_type_data_array(data: xr.DataArray) -> None:
...
+    xarray.DataArray
+        An xarray DataArray containing the computed path length.
+        Will have the same dimensions as the input data, except for ``time``
+        and ``space`` which will be removed.
+
+    """
+    # Skip first displacement segment (always 0) to not mess up the scaling
+    displacement = compute_displacement(data).isel(time=slice(1, None))
+    # count number of valid displacement segments per point track
+    valid_segments = (~displacement.isnull()).any(dim="space").sum(dim="time")
```
[the remark this answers is not in the document]
Comment (niksirbi) [reply movement/analysis/kinematics.py:572]: You are actually right, I want to consider a point invalid if any of its 2 spatial coordinates is NaN. What I missed is that that because of the `~`, I flip the conditional, so `any` must become `all`.

File: movement/analysis/kinematics.py [hunk 4]
```diff
@@ -343,3 +444,180 @@ def _validate_type_data_array(data: xr.DataArray) -> None:
...
+            ValueError,
+            "nan_warn_threshold must be a number between 0 and 1.",
+        )
+
+    n_nans = data.isnull().any(dim="space").sum(dim="time")
+    data_to_warn_about = data.where(
+        n_nans > data.sizes["time"] * nan_warn_threshold, drop=True
+    )
+    if len(data_to_warn_about) > 0:
+        log_warning(
```
[the remark this answers is not in the document]
Comment (niksirbi) [reply movement/analysis/kinematics.py:537]: hmm, it feels like we should be printing warning on the console. Should we (in general) configure the logger such that W are printed on the console as well as being logged?

File: movement/analysis/kinematics.py [hunk 5]
```diff
@@ -343,3 +444,180 @@ def _validate_type_data_array(data: xr.DataArray) -> None:
...
+            "The result may be unreliable for point tracks with many "
+            "missing values. The following tracks have more than "
+            f"{nan_warn_threshold * 100:.3} %) NaN values:",
+        )
+        report_nan_values(data_to_warn_about)
```
[the remark this answers is not in the document]
Comment (niksirbi) [reply movement/analysis/kinematics.py:542]: Good catch!

File: movement/analysis/kinematics.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (niksirbi) [reply movement/analysis/kinematics.py:537]: I opened an [issue](https://github.com/neuroinformatics-unit/movement/issues/337), to address what I *think* is the root cause of this problem.

File: movement/analysis/kinematics.py [hunk 6]
```diff
@@ -171,6 +172,106 @@ def compute_time_derivative(data: xr.DataArray, order: int) -> xr.DataArray:
...
+
+    Choosing ``nan_policy="scale"`` will adjust the path length based on the
+    the proportion of valid segments per point track. For example, if only
+    80% of segments are present, the path length will be computed based on
+    these and the result will be divided by 0.8. This approach assumes
+    that motion dynamics are similar across present and missing time
+    segments, which may not be the case.
+
+    """
+    _validate_start_stop_times(data, start, stop)
```
[the remark this answers is not in the document]
Comment (niksirbi) [reply movement/analysis/kinematics.py:258]: I agree we should rely on `slice` to do its thing and not add extra validation here. 
I've essentially removed start-stop time checks entirely, except for one case (see below).

xarray `slices`, make some specific choices:
- if start/stop time are out of the range, it only returns what's in the data's range (which is sensible)
- if start/stop are incorrect types (e.g. string), it does not error, and it silently returns an array with time dimension of size 0.

I find the 2nd choice weird and it messes up the downstream computation of path length (it will error out, with an error message that's not so easy to parse).

To account for this (time dim with size 0) and also for the edge case where the time dim has size 1 (only 1 frame selected), I now throw a somewhat informative error if the size of the time dim < 2 (in which case no meaningful path length can be computed).

I kept the test across time ranges, but I adapted it accordingly. It's useful to keep this test, to remind ourselves of the above behaviours.

File: tests/test_unit/test_kinematics.py [hunk 7]
```diff
@@ -186,6 +204,223 @@ def test_approximate_derivative_with_invalid_order(order):
...
+            pytest.raises(
+                TypeError, match="Expected a numeric value for stop"
+            ),
+        ),
+    ],
+)
+@pytest.mark.parametrize(
+    "nan_policy",
+    ["drop", "scale"],  # results should be same for both here
+)
```
[the remark this answers is not in the document]
Comment (niksirbi) [reply tests/test_unit/test_kinematics.py:262]: agreed

File: tests/test_unit/test_kinematics.py [hunk 8]
```diff
@@ -186,6 +204,223 @@ def test_approximate_derivative_with_invalid_order(order):
...
+    "nan_policy, expected_path_lengths_id_1, expected_exception",
+    [
+        (
+            "drop",
+            {
+                # 9 segments - 1 missing on edge
+                "centroid": np.sqrt(2) * 8,
+                # missing mid frames should have no effect
+                "left": np.sqrt(2) * 9,
+                "right": np.nan,  # all frames missing
```
[the remark this answers is not in the document]
Comment (niksirbi) [reply tests/test_unit/test_kinematics.py:302]: I've decided to adopt you suggestion. The test function's docstring is sufficient to explain what this test expects.

File: movement/analysis/kinematics.py [hunk 9]
```diff
@@ -343,3 +444,180 @@ def _validate_type_data_array(data: xr.DataArray) -> None:
...
+        track_data = stacked_data.sel(tracks=track_name, drop=True).dropna(
+            dim="time", how="any"
+        )
+        # Compute path length for current point track
+        # and store it in the result array
+        target_loc = {k: v for k, v in zip(dims_to_stack, track_name)}
+        path_length.loc[target_loc] = compute_norm(
+            compute_displacement(track_data)
+        ).sum(dim="time", min_count=1)  # returns NaN if no valid segment
+    return path_length
```
[the remark this answers is not in the document]
Comment (niksirbi) [reply movement/analysis/kinematics.py:574]: I like this, and it's much more concise. In fact, groupby is not needed at all I think. The following works just the same:
```python
return compute_norm(
        compute_displacement(data.ffill(dim="time")).isel(time=slice(1, None))
    ).sum(dim="time", min_count=1)
```

I had to thing why 'ffill' achieves the same thing as 'drop', and I think I've got it. When we apply 'ffill', it's as if the point was stationary throughout the missing segment (holding the last known valid position), and then abruptly moved to the next valid position. The displacement vector from the last ffilled-position to the next valid position will correspond to a straight-line motion, so effectively the same as for the 'drop' case.

I will rename the 'drop' policy to 'ffill', and adjust the docstrings accordingly.

**Comment by niksirbi** [comment]
Thanks for the thorough review @lochhh!

- I've accepted your suggestions for docstring wording.
- I dropped the custom start/stop time checks, relying mostly on `xarray.sel` to handle this. That said, I did choose to still catch one specific edge case (see relevant comment)
- I adopted and further simplified your `ffill` approach (instead of `drop`). It achieves the same result and is much more concise!
- Simplified test parametrisation, following your suggestions.
- Rebased on `main` and dealt with merge conflicts.

**Review by lochhh — APPROVED** [review]
Brilliant! 🚀 Thanks for addressing all the comments!
